gdraw/ggdkdraw.c

```diff
@@ -785,30 +789,44 @@ static bool _GGDKDraw_FilterByModal(GdkEvent *event, GGDKWindow gw) {
             break;
     }
 
-    while (gww != gw->display->groot) {
-        if (gww->transient_childs->len > 0) {
-            bool has_modal = false;
-            for (guint i = 0; i < gww->transient_childs->len; i++) {
-                GGDKWindow ow = (GGDKWindow)gww->transient_childs->pdata[i];
-                if (ow->restrict_input_to_me) {
-                    has_modal = true;
-                    break;
-                }
-            }
-            if (has_modal) {
-                break;
-            }
-        }
-        gww = gww->parent;
-    }
+    GGDKWindow gww = gw;
+    GPtrArray *stack = gw->display->transient_stack;
 
-    if (gww == gw->display->groot) {
-        return false;
+#ifdef FONTFORGE_CAN_USE_GTK_BRIDGE
+    if ( gtkb_Grabbed(gww->display->gtkb_state) ) {
+	// If gtk has an active grab, push the (already filtered)
+	// event up to it.
+	gtkb_ProcessEvent(gww->display->gtkb_state, event);
+	// Keep the "beep" (XXX?)
+	gww = NULL;
     }
+#endif
+
+    if ( gww && gw->display->restrict_count == 0 )
```

Comment:
I'm not 100% on the need for `restrict_count`, vs. just checking the length of the transient stack/list.

Reply:
Clearly not everything in the transient list is meant to have `restrict_input_to_me` set, given that the code checks the flag. And it doesn't seem like the bottom would be guaranteed to have it.

Reply:
My first response was a little terse. 

Suppose you right-click on a CharView to bring up the popup menu. I believe that the menu "window" is a transient for the CharView but doesn't set `restrict_input_to_me` (you can use other windows while that menu is "popped"). Now suppose you use that menu to set an anchor point. GDraw may not show the menu when the subsequent dialog is displayed, but it's "there" and will only get cleaned up later. I *believe* it's in the stack still as well, but I'm not sure. Anyway, there are other plausible scenarios in which you can have: a) a base window, b) a transient non-restricting window, and then c) a "restricting" or modal transient window. 

So what GTK would call a "grab" -- where you can't use other windows -- corresponds to whether there is a displayed window that sets `restrict_count_to_me`. Once you have one of these, any further transients *setting that flag* should be transients for that window (unless some other event, like a python call, brings up another modal window somehow) , but the number is as arbitrary as the number of modals that can bring up further modals (in practice, probably 3?). 

So to determine whether there is presently a modal grab somewhere, you can either go through the full transient list or track the number of `restrict_input_to_me` flags among those transients. And if you want a *signal* when the status changes, you need to do one of those two things whenever the list contents change. Since there was already optimization around adding and removing elements from the list, I went with tracking the value rather than iterating every time.

Reply:
Ok, I've spent way too long looking at this. In part, a transient window *implies* modality (or at least modality for the window it was transient for).

Wrt. the case you mention, the menu is actually destroyed before the new window is created, and it is also marked with the restrict_input_to_me flag.

Anyway, I've kept it, but I've moved everything into the GGDKDrawSetTransientFor call, so if you need to track that transition, do it in there. (there was also a potential bug of the count being incremented too much if say GGDKDrawSetTransientFor was called multiple times on the same window (where ow!=null), although I guess this isn't really a valid thing to do.

I've also added a check that setting transients can only be done on toplevel (and not child) windows.

Reply:
That works. The new review looks good to me.